### PR TITLE
net: buf: Redesign with split data and meta-data

### PR DESCRIPTION
**Pull request by jhedberg**
This is work in progress, but published as a PR so others can collaborate on it.

Related to https://github.com/zephyrproject-rtos/zephyr/issues/5378.

### CODE REVIEW

**Comment by pfalcon** [comment]
This is related to https://github.com/zephyrproject-rtos/zephyr/issues/5378 . I'll also edit description to put this ref there.

**Review by pfalcon — DISMISSED** [review]
This patchset introduces ability to allocate variable-size fragments, which leads to a concern of internal memory fragmentation in a pool which holds them, discussed in https://github.com/zephyrproject-rtos/zephyr/issues/5378#issuecomment-356626903  and following comments.

I consider that a bit concern, so would like to ask for detailed motivation and benefit/risk analysis. Let me mark this as -1 until then.

I would also suggest to split the changes which are given in the subject of this PR ("split data and meta-data"), from changes which introduce variable-size allocation.

Thanks.

**Comment by jhedberg** [comment]
> This patchset introduces ability to allocate variable-size fragments, which leads to a concern of internal memory fragmentation in a pool which holds them, discussed in #5378 (comment) and following comments.

@pfalcon have you noticed that the the API now exposes the ability to have custom allocators, which don't need to use variable-size fragments? I introduced this specifically because of your earlier stated concerns. In fact, all conversions to the new API I've done so far (tests/net/buf is an exception) use NET_BUF_POOL_FIXED_DEFINE which uses similar fixed-sized data blocks like the old implementation used.

**Comment by jhedberg** [comment]
Following up on my previous comment: I haven't yet had the chance to create a good overview documentation of allocator handling, but the short story is this: the user is free to use their own `struct net_buf_data_alloc` implementation when defining buffer pools. There are three "standard" provided allocators if the user doesn't want to define its own: NET_BUF_POOL_HEAP_DEFINE which uses the OS heap, NET_BUF_POOL_VAR_DEFINE which defines a variable sized allocator using k_mem_pool (this is all that there was in my first draft of this feature), and NET_BUF_POOL_FIXED_DEFINE, which is the closes to the old implementation, i.e. fixed size chunks.

**Comment by pfalcon** [comment]
> have you noticed...  In fact, all conversions to the new API I've done so far...

Indeed, what I noticed form my quick look/grep is that you add macros to define var-sized pool, and define one, but it's not really used (beyond tests).

Still, I'd like to suggest to go in a top-down manner, starting with requirements for having var-sized fragments, then going to the code which implements them. And suggestion to split var-sized frag introduction still holds.

(I understand that you may be busy with other things, and don't expect immediate results, and I'm working on other things, so just trying to record the feedback I have before I forget it. Thanks.)

**Comment by jhedberg** [comment]
> Still, I'd like to suggest to go in a top-down manner, starting with requirements for having var-sized fragments, then going to the code which implements them. And suggestion to split var-sized frag introduction still holds.

I've already done this analysis for the Bluetooth side, and at least there it will likely be more optimal to use variable sized buffers for some use cases, such as HCI events which can vary greatly in their size.

I'd expect the choice of data allocator to be done based on the needs of each net_buf_pool, i.e. after profiling and comparing which allocation strategy works the best. The simplest way to approach this would be to convert all existing users to NET_BUF_POOL_FIXED_DEFINE, since that gives you roughly the same behavior as before, and then go through on a per-pool basis of what may make sense to convert to a different kind of allocator.

File: include/net/buf.h
Comment (pfalcon) [inline include/net/buf.h]: This seems like a regression. Previously, fixed-size buf pool supported blocking on alloc. It's also the baseline pool implementation. By both reasons it should support timeouts IMHO.

File: subsys/net/Kconfig
Comment (pfalcon) [inline subsys/net/Kconfig]: Why it can't be 4? Why can't be 1 (just some flag)? Why default is 8?

File: subsys/net/buf.c
Comment (pfalcon) [inline subsys/net/buf.c]: This sounds a bit strange. If buffer's *data* (vs buffer itself) can be ref counted at all, then it should be ref-counted regardless of the underlying implementation.

File: subsys/net/Kconfig
Comment (jhedberg) [reply subsys/net/Kconfig]: You're right. This could even be 0. However then we either need Kconfig dependencies or build asserts to make sure the networking and Bluetooth stacks don't get built with anything smaller than what they need (e.g. Bluetooth needs 8)

File: include/net/buf.h
Comment (jhedberg) [reply include/net/buf.h]: The NET_BUF_POOL_FIXED pool type creates a data array with one chunk for each net_buf in the pool. Because of this there's no need to block on the data allocation since there's always guaranteed to be a data block available. The net_buf allocation itself still uses the given timeout like before.

File: include/net/buf.h
Comment (jhedberg) [reply include/net/buf.h]: Will fix. Thanks.

File: include/net/buf.h
Comment (jhedberg) [reply include/net/buf.h]: Yep. Copy-paste issue probably. Will fix. Thanks.

File: include/net/buf.h
Comment (jhedberg) [reply include/net/buf.h]: Yep, there should be a note that not all pools are able to block on the data allocation part. All of them can block on the net_buf allocation part however.

File: include/net/buf.h
Comment (pfalcon) [reply include/net/buf.h]: > The net_buf allocation itself still uses the given timeout like before.

This is good point which needs to be emphasized - that API-level net_buf_alloc() operation actually consists of net_buf header allocation and actual data backing store allocation, both of which can block (or not block) independently. I'm not sure where's the best place to describe that though. Firsts starting point is having this point in review comments indeed.

File: subsys/net/buf.c
Comment (jhedberg) [reply subsys/net/buf.c]: Sure, not completely sure about a struct for a single u8_t, but I'll consider it.

File: subsys/net/buf.c
Comment (jhedberg) [reply subsys/net/buf.c]: The net_buf data pointer has never carried with it a promise that it follows a certain alignment. But you're right that some users of net_buf may benefit from it. The simplest solution (not requiring any change to the net_buf implementation) is for the user to use net_buf_reserve() to increment the data pointer until it falls on the desired alignment boundary. Another option is to put the ref count at the end of the data instead of the beginning.

File: subsys/net/buf.c
Comment (jhedberg) [reply subsys/net/buf.c]: The data allocation API gives the option for implementations that do not support refcounting. For now I opted for the simplest possible implementation for the fixed pool, where reference counting is not supported. This means that currently when you do a net_buf_clone() on such a buffer a new data portion is allocated and the data is copied. Adding ref-counting support is not too hard: we'd need a bitmap to track which data blocks are allocated and look up free ones using e.g.  __builtin_ffs().

File: include/net/buf.h
Comment (jhedberg) [reply include/net/buf.h]: Another issue to consider is how strictly we want to adhere to the exact timeout that's given when both net_buf and data allocation may block. Say the user gives a 10 second timeout to net_buf_alloc() and gets a net_buf in 5 seconds. The current implementation would potentially block for another 10 seconds when waiting for the data, instead of subtracting the time waited from the original timeout.

**Review by dbkinder — APPROVED** [review]
+1 for doc edits

File: include/net/buf.h
Comment (jhedberg) [reply include/net/buf.h]: I've updated the PR with another patch (the second one) which now subtracts the time already spent with the net_buf allocation when proceeding to allocating the data portion.

Comment (jhedberg) [reply include/net/buf.h]: Hmm, github doesn't show it as the second one (even though it is in the commit history. Anyway, it's "net: buf: Fix passing too big timeout to data_alloc() "

Comment (pfalcon) [reply include/net/buf.h]: > Another issue to consider is how strictly we want to adhere to the exact timeout

Yeah, I thought about that too. My idea was to just document that maximum wait can be 2 * timeout specified by user, because we have bigger problems to resolve for now. Well, great that you went to resolve it in a better way, that particular patch looks clean. Thanks!

File: subsys/net/buf.c
Comment (pfalcon) [reply subsys/net/buf.c]: > The data allocation API gives the option for implementations that do not support refcounting.

So, one problem with that is that there're just too many options (that's a general problem with this PR as a whole, but let's leave that for "summary" review). For example, I'm not entirely sure is that choice would leak into user-level API, or would be terminated at functions like net_buf_clone().

> For now I opted for the simplest possible implementation for the fixed pool

I don't agree with such a choice either. The fixed-size fragments is what we had before, and should be the baseline of this functionality.Then, it should be converted to offer the same functionality as it had before. And I believe that this refactor noticeably changes its behavior, which I'd recommend against.

Comment (jhedberg) [reply subsys/net/buf.c]: I'm not completely following. Are you against the fixed-size fragments with or without refcounting? Previously there was no refcounting, so right now this PR maintains the old behavior.

As for too many options, previously you objected to there only being a single option (k_mem_pool), and now that there can be custom allocators there are too many? I guess we have to find some balance mid-way :)

**Comment by pfalcon** [comment]
@jhedberg : I appreciate the discussion of this PR. It's scope and intent is now more clear. I'd like to provide some summary of my review.

> I've already done this analysis for the Bluetooth side, and at least there it will likely be more optimal to use variable sized buffers for some use cases

That's actually very useful information, along the lines of "top-down approach" and requirements specification for the refactor. So, let me try to formulate my first point as:

1. The *leading* stakeholder behind these changes appears to be the Bluetooth subsystem. I emphasize *leading*, because I know that @tbursztyka waited and welcomes these changes (and I can agree they're for good either). Still, there's a question of the scope of the change and process of applying them. And for example, for variable-sized buffers introduction, I didn't hear information that they would be beneficial for IP stack. So, I guess it may be fair to say that changes may be rationed and staged to not cause too much disruption in the other subsystem.


2. The other opinion I'd like to share is that there're too many changes in this PR. Like, the PR overall, and some commits in particular. I'd like to extend my suggestion to split some commits into more focused and fine-grained ones. A good example is "net: buf: Redesigned API with split data and meta-data". It's named like that, but also removes an implementation of the older fixed-sized buffers and replaces it with 3 new, none of which appear to be fully functionally equivalent to the older implementation. (Maybe one of them is, but due to massive changes, it's hard to reason about that.)

> you noticed that the API now exposes the ability to have custom allocators, which don't need to use variable-size fragments? I introduced this specifically because of your earlier stated concerns.

3. I appreciate the above, and I'm all for different choices given different trade-offs, and for configurability. So, it's nice that we're going to have 3 concurrent buffer implementations. But that leads to a problem: how do we test all that? I think that's worth having an "official" response to that. From my side, I'd think that the scalable approach is to make new fixed-size fragment buffers implementation to be as close functional equivalent of the previous implementation as possible.

4. Despite 3 concurrent implementations of buffer allocators being introduced, I don't see another usecase being addressed - of dealing with constant, static data (coming from ROM for example). Note that this usecase is "orthogonal" to the selection of 3 allocator types above: an application selects one of the allocators above, and with that, should be able to either allocate RAM-based data buffer (using one of 3 allocation policies) **or** allocate ROM-based buffer. Finding a good API for that is a separate (and important) task, e.g. from a user PoV it's beneficial to make it as easy as passing a flag to an existing buffer alloc function (net_buf_alloc() in our case).


I provide this feedback with the outlook that 1.11 feature freeze is 1 week away, and it would be nice get the initial, concerns-free part(s) of the refactor into it. (And also with the understanding that you're the part of the core team, and define what goes into the release yourselves. So, my opinion is neither deciding nor blocking.) Thanks!

File: subsys/net/buf.c
Comment (pfalcon) [reply subsys/net/buf.c]: > Are you against the fixed-size fragments with or without refcounting? Previously there was no refcounting, so right now this PR maintains the old behavior.

Let me put it this way: does net_buf_clone() *and* net_pkt_clone() with new fixed-size impl behave the same way as they did with the old? If so, ok, but this fact isn't immediately obvious (and as I mentioned, to amount of changes, it's a bit hard to focus on review one topic completely (would take a long time to review entire PR)).

> previously you objected

Well, it's not that I "object", but I try to provide multi-faceted review, pointing out various concerns. As I wrote in the summary, I don't "object", I'm all for, but think you need to give an idea how these changes would be tested. Would be build each and every sample and test with all of 3 options in sanitycheck? Something else?

File: subsys/net/buf.c
Comment (pfalcon) [inline subsys/net/buf.c]: I don't think that's useful behavior. Let's consider an algorithm to prepare a pkt for sending:

1. Take interface MTU as a basic constraint on the pkt size.
2. Apply additional constraints: connection and protocol specific (e.g. path MTU, TCP's MSS, etc.).
3. Thus, a max pkt size is found.
4. Add protocol headers, decrease the size.
5. Whatever is lesser, input user data size, or remaining pkt size, allocate a data fragment buffer of that size. (Note: given different allocator implementation, a client knows *nothing* about max size a data fragment allocator can handle).
6. Fill in receive data fragment, subtract fragment size from remaining pkt size, if not zero, repeat from 5.


Thus, data buffer allocator would need:

1. Be ready to accept request of arbitrary allocation size.
2. Try to satisfy some of that size (even if smaller). 
3. Allow client to query the size of the allocated buffer, (for step 6 above).

File: subsys/net/buf.c
Comment (jhedberg) [reply subsys/net/buf.c]: I can't comment for net_pkt since I'm not familiar with that code, but for net_buf the semantics of buffer cloning remain at the moment the same with NET_BUF_POOL_FIXED.

File: subsys/net/buf.c
Comment (jhedberg) [reply subsys/net/buf.c]: This analysis makes very much sense to me. There's already buf->size which indicates the maximum payload size, so what's left to do is to document net_buf_alloc() that it may return a smaller than requested payload if the allocator wasn't able to fulfill the exact request. FWIW, on the Bluetooth side it would in many cases constitute a bug to try to do such an allocation since the pool definition and allocation code are quite tightly coupled. This is the reason I initially implemented this by throwing an error.

Comment (jhedberg) [reply subsys/net/buf.c]: I've now updated the PR so that the alloc callback is able to allocate less than requested, and the FIXED pool allocator will not fail anymore, but allocate at most the fixed block size.

File: subsys/net/buf.c
Comment (pfalcon) [inline subsys/net/buf.c]: Should assign to *size?

File: subsys/net/buf.c
Comment (pfalcon) [inline subsys/net/buf.c]: Should assign to *size?

File: subsys/net/buf.c
Comment (pfalcon) [reply subsys/net/buf.c]: > I've now updated the PR so that the alloc callback is able to allocate less than requested

Thanks, much appreciated!

File: subsys/net/buf.c
Comment (jhedberg) [reply subsys/net/buf.c]: I'm not changing anything related to the allocation of the net_buf structs themselves. I.e. they're still managed through a k_lifo like before. It's only the data allocation part that has extra structs etc.

File: subsys/net/buf.c
Comment (jhedberg) [reply subsys/net/buf.c]: No, since this size is what gets set to buf->size, i.e. it should not take into account internal metadata such as the refcount.

File: subsys/net/buf.c
Comment (jhedberg) [reply subsys/net/buf.c]: No, since this size is what gets set to buf->size, i.e. it should not take into account internal metadata such as the refcount.

**Comment by jhedberg** [comment]
I've updated the PR for an incremental approach to the API change. The new allocator (which takes a length parameter) is now called net_buf_alloc_len, so it doesn't conflict with the old net_buf_alloc API. I've also renamed net_buf_alloc to net_buf_alloc_fixed, and made an alias of net_buf_alloc to point at that. The old NET_BUF_POOL_DEFINE() macro is now still around, and maps to NET_BUF_POOL_FIXED_DEFINE(). The only thing it does with the user_data_size parameter is to have a build assert that it's bigger than the new Kconfig value. I've still got the various patches split up here, so you can see the work, but once everyone is fine with this approach I'll squash the patches to two:

 - New API introduction and a few code updates to remove usage of old pool members
- Update of tests/net/buf to use the new API

Once this is merged we can then start looking at converting the code incrementally.

**Comment by pfalcon** [comment]
> I've still got the various patches split up here, so you can see the work, but once everyone is fine with this approach I'll squash the patches to two:

Thanks for this update. But the lines above go in the opposite direction to my suggestion of splitting up this patchset into more logically-separate patches, then perhaps into separate PRs, and look forward to merging the first one of them in the 1.11 timeframe, to not risk the whole refactor going into 1.13 or something. 

To avoid any ambiguity, what I proposed is to start with this change:

~~~
 struct net_buf_simple {
...
-	u8_t __buf[0] __net_buf_align;
+	u8_t *__buf;
 }
~~~

Then propagate that change to other parts of code, and that would be the initial patch, which, from my PoV as a contributing engineer (not a maintainer), would be safe to merge into 1.11.

@jhedberg, it would be nice to get explicit confirmation that you don't like this plan, and ideally, reasons why. Thanks. (For reference, "split up patches" is a frequent request in Zephyr review comments overall.)

**Comment by jhedberg** [comment]
Hi @pfalcon 

The way I understood "ability to split up into patches" was that we should be able to update one module at a time in the tree to the new API. That's exactly what I've now enabled. The PR touches almost no code outside of the net_buf module, i.e. the remaining changes can be done incrementally in small patches as part of future pull requests. The only stuff outside of the net_buf module I had to touch was due to the removal of the buf_size and user_data_size members of the net_buf_pool struct.

I'm happy with this approach, as are @tbursztyka and @jukkar

What I'd now welcome from anyone with the time to do so, is some manual testing of this PR. I've tested it with as many Bluetooth use cases I could, and it seems to work fine, but the networking side is another matter.

**Comment by pfalcon** [comment]
> The way I understood "ability to split up into patches" was that we should be able to update one module at a time in the tree to the new API.

I got a feeling that there may be confusion about that, yes. What I meant from my first review comment is splitting the changes into logically separate, self-contained parts. For example, splitting metadata and data parts of net_buf's (making metadata contain a pointer to data) is one standalone part, while introduction of allocators with different allocating policies is another self-contained change. Structuring the changes in such way largely simplifies and makes more robust the review process, and allows to perform more agile project planning (e.g. to react to criticism (by @nashif, at the December network forum meeting) that the overall refactor plan asks for too much time, and apparently conflicts with the LTS release plans)).

All in all, I writing this to disambiguate my position, why I gave -1 mark to this PR, and feel that concerns raised weren't addressed, so my mark persists. And just like at the very beginning, I'm sure that my mark won't lead to any issues/delays with merging this patch, beyond those rooted in its existing complexity, as there will be enough parties to override my mark. Thanks.

**Comment by jhedberg** [comment]
@pfalcon there's no need to reiterate the benefits of favoring small logical changes to big ones, it's something I'm very much familiar with and not something I disagree with. However, in this case I'd need to rewrite the fixed pool allocator without the custom allocator support, only to remove the code in a later patch. What I can do is split the _VAR and _HEAP allocators into a separate patch. But that's the extent to which I think further splitting makes sense.

Regarding the confusion with the splitting. About 90% of my original changes were in the single patch that went and modified various parts of the tree to use the new API. That was the natural thing my attention went to when you requested some splitting to occur. That 90% is now completely removed, and we're left only with the net_buf module changes themselves.

**Comment by jukkar** [comment]
To my eyes the patches look ok, are properly split, and they do not modify existing behavior (just need to figure out what goes wrong with coap test). It looks like this will miss 1.11 release if we do not find consensus about this PR.

**Comment by tbursztyka** [comment]
I don't see any logical issue with the patch amount and ordering. (besides, maybe the subsys/net/Kconfig removal of depends on towards using if, but that's really super minor).

Actually the way it's done simplifies a lot the further PRs (moving net stack to it, bt stack etc...)

**Comment by jhedberg** [comment]
I've now done the split into the three separate patches, as envisioned. I still need to write proper commit messages, and double check if something is needed in the documentation (that said, @dbkinder already approved ;)

**Comment by pfalcon** [comment]
@jhedberg:

> there's no need to reiterate the benefits of favoring small logical changes to big ones

I indeed repeated myself there, but given the situation that we all, myself included, agree that this is useful refactor, and I only wished it started to stream into the mainline, it may sound "counter-intuitive" that I give -1 to it. Given that this is public review, I err on the side of repeating myself to extricate the situation: I'm all for the factor, but don't agree with the specific process of how changes are being made. (I for one wouldn't like to be in the same situation - when on words we agree, but my patch gets -1 without clear explanation/a way out of it ;-) ).

> it's something I'm very much familiar with and not something I disagree with

Again, given that this is a public review, not just communication between a patch developer and a reviewer, I'm left with the little choice as to verbosely explain my position, because otherwise it may look to 3rd parties as I request some changes on a whim, without even trying to explain what would be the benefits of it. But bottom line: if we all familiar with that, then maybe one of these days we should try to follow that process and optimize for reviewability, not trying to get convenient shortcuts around it ;-).

> However, in this case I'd need to rewrite the fixed pool allocator without the custom allocator support, only to remove the code in a later patch.

I very much appreciate this explanation, it's something which I hoped for. And it's hard to argue with that of course, especially, if for you, as the author, it's crystal clear that that change will need to be done in exactly the way you do, while a reviewers says "I can't clearly see that so far, maybe after the changes are separated" (because that's literally what I say). The only thing I can say still: "if only it was like that from the beginning.". And maybe if we eventually would get more of such split-in-the beginning patches, all this "workflow methodology" discussion wasn't just waste of time.

And btw, I keep up this discussion still not on a whim, but as reaction to issues like a) https://github.com/zephyrproject-rtos/zephyr/issues/5484, https://github.com/zephyrproject-rtos/zephyr/issues/5857, https://github.com/zephyrproject-rtos/zephyr/issues/5952, etc.; b) "news" that we're going to "adopt" MISRA and other stuff. MISRA would require us to much more acrobatics (as it's full of ~~nonsense~~ things like "a function should have only one exit"), but isn't going to help if there're hard-to-review changes (due to their size and making many things into one commit).

**Comment by pfalcon** [comment]
> It looks like this will miss 1.11 release if we do not find consensus about this PR.

Ok. Then let my -1 signify the fact that this patch as a whole isn't suitable for 1.11. I'll then know that for 1.11 it's better to concentrate on other things, and set aside some good time later to review this patch as a whole. (I only will be happy if other people will beat me on that.)

**Comment by tbursztyka** [comment]
@pfalcon 
> Ok. Then let my -1 signify the fact that this patch as a whole isn't suitable for 1.11. I'll then know that for 1.11 it's better to concentrate on other things, and set aside some good time later to review this patch as a whole. (I only will be happy if other people will beat me on that.)

So we had a TLC meeting last week - you were present I think - telling the plan is to provide net_buf split asap (so 1.11) and you are already killing the plan just by... taste issue? ^^

Seriousyl, propagating small changes after small changes is, in this case, not really welcomed. It will surely create tiny patch on net_buf side easier to review, but big ones on the subsystems using it (name or parameter change etc...). And that repeatedly. It will take some time before we get to the real meat here: all the possibilities it offers on optimizing memory and performances.

**Comment by pfalcon** [comment]
> So we had a TLC meeting last week - you were present I think - telling the plan is to provide net_buf split asap (so 1.11) 

That's not what I heard (Zoom.us communication issues + always can remember that I'm not a native speaker). What I heard is that the plan is to proceed with the socket quest immediately, not blocked on net_buf refactor or anything, while being ready to adjust for it *when* it gets ready for the mainline, and that "when" didn't include 1.11. (My personal idea was that @jhedberg is busy with BT for 1.11, and this is on backburner. What I was trying to do is to "motivate" him to split of the initial part to go into 1.11). 

> and you are already killing the plan just by... taste issue? ^^

As I mentioned, I don't think that my meager -1 can stop the maintainers from doing anything, for example merge this 2 days before the release ;-). I dismissed my review.

> And that repeatedly.  It will take some time before we get to the real meat

In my view, that's good. I gave elsewhere an example that it's better to submit a smaller patch every other day than wait a month to submit 10 patches and that wait for review. (Yes, sometimes that's not possible, but if it is, it should be the preferred way. YMMV)

**Comment by jhedberg** [comment]
tests/net/lib/coap is crashing because net_frag_read_byte() gets called with a net_buf that has ref_count 0 and data NULL. This is now caught since net_buf_unref will set buf->data to NULL when it's freed. The issue happens in the test_block_size() test, when coap_next_block() is called after net_pkt_unref() has been called. @rveerama1 could you look at this? @jukkar told me you should know about this test case.

**Comment by rveerama1** [comment]
@jhedberg https://github.com/zephyrproject-rtos/zephyr/pull/6019 should fix the issue.

File: include/net/buf.h
Comment (tbursztyka) [inline include/net/buf.h]: remove "and flags". looks like c/p issue, from net_buf_reset

File: subsys/net/buf.c
Comment (tbursztyka) [inline subsys/net/buf.c]: Isn't it so that if we clone: we expect to clone all the data, so size must be the same as buf->size. If not, cloning failed.

**Review by tbursztyka — DISMISSED** [review]

**Comment by jhedberg** [comment]
@tbursztyka your two comments have now been addressed

**Review by tbursztyka — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**
